File: example/interfaces.go
// Copyright 2014 The GiterLab Authors. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

// example for ots2
package main

import (
	"fmt"
	"os"
	"time"

	ots2 "github.com/GiterLab/goots"
	. "github.com/GiterLab/goots/otstype"
)

// modify it to yours
const (
	ENDPOINT     = "your_instance_address"
	ACCESSID     = "your_accessid"
	ACCESSKEY    = "your_accesskey"
	INSTANCENAME = "your_instance_name"
)

func main() {
	// set running environment
	ots2.OTSDebugEnable = true
	ots2.OTSLoggerEnable = true
	ots2.OTSErrorPanicMode = true // 默认为开启，如果不喜欢panic则设置此为false

	fmt.Println("Test goots start ...")

	ots_client, err := ots2.New(ENDPOINT, ACCESSID, ACCESSKEY, INSTANCENAME)
	if err != nil {
		fmt.Println(err)
	}

	// delete_table
	ots_err := ots_client.DeleteTable("myTable")
	if ots_err != nil {
		fmt.Println(ots_err)
		// os.Exit(1)
	}
	fmt.Println("表已删除")

	// create_table
	// 注意：OTS是按设置的ReservedThroughput计量收费，即使没有读写也会产生费用。
	table_meta := &OTSTableMeta{
		TableName: "myTable",
		SchemaOfPrimaryKey: OTSSchemaOfPrimaryKey{
			{K: "gid", V: "INTEGER"},
			{K: "uid", V: "INTEGER"},
		},
	}

	reserved_throughput := &OTSReservedThroughput{
		OTSCapacityUnit{0, 0},
	}

	ots_err = ots_client.CreateTable(table_meta, reserved_throughput)
	if ots_err != nil {
		fmt.Println(ots_err)
		os.Exit(1)
	}
	fmt.Println("表已创建")
	fmt.Println("表创建后需要等待一会，才能进行继续操作,否则可能会出现操作失败...")
	for cnt := 5; cnt != 0; cnt-- {
		fmt.Println(cnt)
		time.Sleep(1 * time.Second)
	}

	// list_table
	list_tables, ots_err := ots_client.ListTable()
	if ots_err != nil {
		fmt.Println(ots_err)
		os.Exit(1)
	}
	fmt.Println("表的列表如下：")
	fmt.Println("list_tables:", list_tables.TableNames)

	// update_table
	// 只能在高性能实例实例下测试，不要在容量型实例上测试，否则出错
	//
	// 每次调整操作的间隔应大于10分钟
	// 如果是刚创建表，需要2分钟之后才能调整表的预留读写吞吐量。
	// 注意：OTS是按设置的ReservedThroughput计量收费，即使没有读写也会产生费用。
	//
	// Note:
	// 容量型实例: The value of read capacity unit can only be 0
	//             The value of write capacity unit can only be 0.
	//             Your instance is forbidden to update capacity unit.
	// 高性能实例: at least one of read or write of CapacityUnit is required
	update_reserved_throughput := &OTSReservedThroughput{
		OTSCapacityUnit{0, 0},
	}

	fmt.Println("Need to sleep 2 Minute, be patient...")
	time.Sleep(2 * time.Minute)
	time.Sleep(5 * time.Second)
	update_response, ots_err := ots_client.UpdateTable("myTable", update_reserved_throughput)
	if ots_err != nil {
		fmt.Println(ots_err)
		os.Exit(1)
	}
	fmt.Println("表的预留读吞吐量:", update_response.ReservedThroughputDetails.CapacityUnit.Read)
	fmt.Println("表的预留写吞吐量:", update_response.ReservedThroughputDetails.CapacityUnit.Write)
	fmt.Println("最后一次上调预留读写吞吐量时间:", update_response.ReservedThroughputDetails.LastIncreaseTime)
	fmt.Println("最后一次下调预留读写吞吐量时间:", update_response.ReservedThroughputDetails.LastDecreaseTime)
	fmt.Println("UTC自然日内总的下调预留读写吞吐量次数:", update_response.ReservedThroughputDetails.NumberOfDecreasesToday)

	// describe_table
	describe_response, ots_err := ots_client.DescribeTable("myTable")
	if ots_err != nil {
		fmt.Println(ots_err)
		os.Exit(1)
	}
	fmt.Println("表的名称:", describe_response.TableMeta.TableName)
	fmt.Println("表的主键:", describe_response.TableMeta.SchemaOfPrimaryKey)
	fmt.Println("表的预留读吞吐量:", describe_response.ReservedThroughputDetails.CapacityUnit.Read)
	fmt.Println("表的预留写吞吐量:", describe_response.ReservedThroughputDetails.CapacityUnit.Write)
	fmt.Println("最后一次上调预留读写吞吐量时间:", describe_response.ReservedThroughputDetails.LastIncreaseTime)
	fmt.Println("最后一次下调预留读写吞吐量时间:", describe_response.ReservedThroughputDetails.LastDecreaseTime)
	fmt.Println("UTC自然日内总的下调预留读写吞吐量次数:", describe_response.ReservedThroughputDetails.NumberOfDecreasesToday)

	// put_row
	primary_key := &OTSPrimaryKey{
		"gid": 1,
		"uid": 101,
	}
	attribute_columns := &OTSAttribute{
		"name":    "张三",
		"mobile":  111111111,
		"address": "中国A地",
		"age":     20,
	}
	condition := OTSCondition_EXPECT_NOT_EXIST
	put_row_response, ots_err := ots_client.PutRow("myTable", condition, primary_key, attribute_columns)
	if ots_err != nil {
		fmt.Println(ots_err)
		os.Exit(1)
	}
	fmt.Println("成功插入数据，消耗的写CapacityUnit为:", put_row_response.GetWriteConsumed())

	// get_row
	primary_key = &OTSPrimaryKey{
		"gid": 1,
		"uid": 101,
	}
	columns_to_get := &OTSColumnsToGet{
		"name", "address", "age",
	}
	// columns_to_get = nil
	// read all
	get_row_response, ots_err := ots_client.GetRow("myTable", primary_key, columns_to_get)
	if ots_err != nil {
		fmt.Println(ots_err)
		os.Exit(1)
	}
	fmt.Println("成功读取数据，消耗的读CapacityUnit为:", get_row_response.GetReadConsumed())
	if get_row_response.Row != nil {
		if attribute_columns := get_row_response.Row.GetAttributeColumns(); attribute_columns != nil {
			fmt.Println("name信息:", attribute_columns.Get("name"))
			fmt.Println("address信息:", attribute_columns.Get("address"))
			fmt.Println("age信息:", attribute_columns.Get("age"))
			fmt.Println("mobile信息:", attribute_columns.Get("mobile"))
		} else {
			fmt.Println("未查询到数据")
		}
	} else {
		fmt.Println("未查询到数据")
	}

	// update_row
	primary_key = &OTSPrimaryKey{
		"gid": 1,
		"uid": 101,
	}
	update_of_attribute_columns := &OTSUpdateOfAttribute{
		OTSOperationType_PUT: OTSColumnsToPut{
			"name":    "张三丰",
			"address": "中国B地",
		},

		OTSOperationType_DELETE: OTSColumnsToDelete{
			"mobile", "age",
		},
	}
	condition = OTSCondition_EXPECT_EXIST
	update_row_response, ots_err := ots_client.UpdateRow("myTable", condition, primary_key, update_of_attribute_columns)
	if ots_err != nil {
		fmt.Println(ots_err)
		os.Exit(1)
	}
	fmt.Println("成功插入数据，消耗的写CapacityUnit为:", update_row_response.GetWriteConsumed())

	// get_row
	primary_key = &OTSPrimaryKey{
		"gid": 1,
		"uid": 101,
	}
	columns_to_get = &OTSColumnsToGet{
		"name", "address", "age",
	}
	columns_to_get = nil // read all
	get_row_response, ots_err = ots_client.GetRow("myTable", primary_key, columns_to_get)
	if ots_err != nil {
		fmt.Println(ots_err)
		os.Exit(1)
	}
	fmt.Println("成功读取数据，消耗的读CapacityUnit为:", get_row_response.GetReadConsumed())
	if get_row_response.Row != nil {
		if attribute_columns := get_row_response.Row.GetAttributeColumns(); attribute_columns != nil {
			fmt.Println("name信息:", attribute_columns.Get("name"))
			fmt.Println("address信息:", attribute_columns.Get("address"))
			fmt.Println("age信息:", attribute_columns.Get("age"))
			fmt.Println("mobile信息:", attribute_columns.Get("mobile"))
		} else {
			fmt.Println("未查询到数据")
		}
	} else {
		fmt.Println("未查询到数据")
	}

	// delete_row
	primary_key = &OTSPrimaryKey{
		"gid": 1,
		"uid": 101,
	}
	condition = OTSCondition_IGNORE
	delete_row_response, ots_err := ots_client.DeleteRow("myTable", condition, primary_key)
	if ots_err != nil {
		fmt.Println(ots_err)
		os.Exit(1)
	}
	fmt.Println("成功删除数据，消耗的写CapacityUnit为:", delete_row_response.GetWriteConsumed())

	// get_row
	primary_key = &OTSPrimaryKey{
		"gid": 1,
		"uid": 101,
	}
	columns_to_get = &OTSColumnsToGet{
		"name", "address", "age",
	}
	columns_to_get = nil // read all
	get_row_response, ots_err = ots_client.GetRow("myTable", primary_key, columns_to_get)
	if ots_err != nil {
		fmt.Println(ots_err)
		os.Exit(1)
	}
	fmt.Println("成功读取数据，消耗的读CapacityUnit为:", get_row_response.GetReadConsumed())
	if get_row_response.Row != nil {
		if attribute_columns := get_row_response.Row.GetAttributeColumns(); attribute_columns != nil {
			fmt.Println("name信息:", attribute_columns.Get("name"))
			fmt.Println("address信息:", attribute_columns.Get("address"))
			fmt.Println("age信息:", attribute_columns.Get("age"))
			fmt.Println("mobile信息:", attribute_columns.Get("mobile"))
		} else {
			fmt.Println("未查询到数据")
		}
	} else {
		fmt.Println("未查询到数据")
	}

	// batch_write_row
	// [1] put row
	put_row_item := OTSPutRowItem{
		Condition: OTSCondition_EXPECT_NOT_EXIST, // OTSCondition_IGNORE
		PrimaryKey: OTSPrimaryKey{
			"gid": 2,
			"uid": 202,
		},
		AttributeColumns: OTSAttribute{
			"name":    "李四",
			"address": "中国某地",
			"age":     20,
		},
	}
	// [2] update_row
	update_row_item := OTSUpdateRowItem{
		Condition: OTSCondition_IGNORE,
		PrimaryKey: OTSPrimaryKey{
			"gid": 3,
			"uid": 303,
		},
		UpdateOfAttributeColumns: OTSUpdateOfAttribute{
			OTSOperationType_PUT: OTSColumnsToPut{
				"name":    "李三",
				"address": "中国某地",
			},
			OTSOperationType_DELETE: OTSColumnsToDelete{
				"mobile", "age",
			},
		},
	}
	// [3] delete_row
	delete_row_item := OTSDeleteRowItem{
		Condition: OTSCondition_IGNORE,
		PrimaryKey: OTSPrimaryKey{
			"gid": 4,
			"uid": 404,
		},
	}
	batch_list_write := &OTSBatchWriteRowRequest{
		{
			TableName: "myTable",
			PutRows: OTSPutRows{
				put_row_item,
			},
			UpdateRows: OTSUpdateRows{
				update_row_item,
			},
			DeleteRows: OTSDeleteRows{
				delete_row_item,
			},
		},
		{
			TableName: "notExistTable",
			PutRows: OTSPutRows{
				put_row_item,
			},
			UpdateRows: OTSUpdateRows{
				update_row_item,
			},
			DeleteRows: OTSDeleteRows{
				delete_row_item,
			},
		},
	}
	batch_write_response, ots_err := ots_client.BatchWriteRow(batch_list_write)
	if ots_err != nil {
		fmt.Println(ots_err)
		os.Exit(1)
	}
	// NOTE: 实际测试如果部分行操作失败，不消耗写CapacityUnit，而不是说明书写的整体失败
	if batch_write_response != nil {
		var succeed_total, failed_total, consumed_write_total int32
		for _, v := range batch_write_response.Tables {
			fmt.Println("操作的表名:", v.TableName)
			fmt.Println("操作 PUT:")
			if len(v.PutRows) != 0 {
				for i1, v1 := range v.PutRows {
					if v1.IsOk {
						succeed_total = succeed_total + 1
						fmt.Println("   --第", i1, "行操作成功, 消耗写CapacityUnit为", v1.Consumed.GetWrite())
						// NOTE: 为什么这里当条件设置为 OTSCondition_IGNORE, 同时这个put的PK值已经存在时
						// 一个put会消耗2个CapacityUnit呢???
						consumed_write_total = consumed_write_total + v1.Consumed.GetWrite()
					} else {
						failed_total = failed_total + 1
						if v1.Consumed == nil {
							fmt.Println("   --第", i1, "行操作失败, 消耗写CapacityUnit为", 0, "ErrorCode:", v1.ErrorCode, "ErrorMessage:", v1.ErrorMessage)
						} else {
							// 实际测试这里不会执行到
							fmt.Println("   --第", i1, "行操作失败, 消耗写CapacityUnit为", v1.Consumed.GetWrite, "ErrorCode:", v1.ErrorCode, "ErrorMessage:", v1.ErrorMessage)
							consumed_write_total = consumed_write_total + v1.Consumed.GetWrite()
						}
					}
				}
			}
			fmt.Println("操作 Update:")
			if len(v.UpdateRows) != 0 {
				for i1, v1 := range v.UpdateRows {
					if v1.IsOk {
						succeed_total = succeed_total + 1
						fmt.Println("   --第", i1, "行操作成功, 消耗写CapacityUnit为", v1.Consumed.GetWrite())
						consumed_write_total = consumed_write_total + v1.Consumed.GetWrite()
					} else {
						failed_total = failed_total + 1
						if v1.Consumed == nil {
							fmt.Println("   --第", i1, "行操作失败, 消耗写CapacityUnit为", 0, "ErrorCode:", v1.ErrorCode, "ErrorMessage:", v1.ErrorMessage)
						} else {
							// 实际测试这里不会执行到
							fmt.Println("   --第", i1, "行操作失败, 消耗写CapacityUnit为", v1.Consumed.GetWrite, "ErrorCode:", v1.ErrorCode, "ErrorMessage:", v1.ErrorMessage)
							consumed_write_total = consumed_write_total + v1.Consumed.GetWrite()
						}
					}
				}
			}
			fmt.Println("操作 Delete:")
			if len(v.DeleteRows) != 0 {
				for i1, v1 := range v.DeleteRows {
					if v1.IsOk {
						succeed_total = succeed_total + 1
						fmt.Println("   --第", i1, "行操作成功, 消耗写CapacityUnit为", v1.Consumed.GetWrite())
						consumed_write_total = consumed_write_total + v1.Consumed.GetWrite()
					} else {
						failed_total = failed_total + 1
						if v1.Consumed == nil {
							fmt.Println("   --第", i1, "行操作失败, 消耗写CapacityUnit为", 0, "ErrorCode:", v1.ErrorCode, "ErrorMessage:", v1.ErrorMessage)

						} else {
							// 实际测试这里不会执行到
							fmt.Println("   --第", i1, "行操作失败, 消耗写CapacityUnit为", v1.Consumed.GetWrite, "ErrorCode:", v1.ErrorCode, "ErrorMessage:", v1.ErrorMessage)
							consumed_write_total = consumed_write_total + v1.Consumed.GetWrite()
						}
					}
				}
			}
		}
		fmt.Printf("本次操作命中 %d 个, 失败 %d 个, 共消耗写CapacityUnit为 %d\n", succeed_total, failed_total, consumed_write_total)
	} else {
		fmt.Println("本次操作都失败，不消耗写CapacityUnit")
	}

	// batch_get_row
	batch_list_get := &OTSBatchGetRowRequest{
		{
			// TableName
			TableName: "myTable",
			// PrimaryKey
			Rows: OTSPrimaryKeyRows{
				{"gid": 1, "uid": 101},
				{"gid": 2, "uid": 202},
				{"gid": 3, "uid": 303},
			},
			// ColumnsToGet
			ColumnsToGet: OTSColumnsToGet{"name", "address", "mobile", "age"},
		},
		{
			// TableName
			TableName: "notExistTable",
			// PrimaryKey
			Rows: OTSPrimaryKeyRows{
				{"gid": 1, "uid": 101},
				{"gid": 2, "uid": 202},
				{"gid": 3, "uid": 303},
			},
			// ColumnsToGet
			ColumnsToGet: OTSColumnsToGet{"name", "address", "mobile", "age"},
		},
	}
	batch_get_response, ots_err := ots_client.BatchGetRow(batch_list_get)
	if ots_err != nil {
		fmt.Println(ots_err)
		os.Exit(1)
	}
	if batch_get_response != nil {
		var succeed_total, failed_total, consumed_write_total int32
		for _, v := range batch_get_response.Tables {
			fmt.Println("操作的表名:", v.TableName)
			for i1, v1 := range v.Rows {
				if v1.IsOk {
					succeed_total = succeed_total + 1
					fmt.Println("   --第", i1, "行操作成功, 消耗读CapacityUnit为", v1.Consumed.GetRead())
					consumed_write_total = consumed_write_total + v1.Consumed.GetRead()
					// print get value
					fmt.Println(v1.Row)
				} else {
					failed_total = failed_total + 1
					if v1.Consumed == nil {
						fmt.Println("   --第", i1, "行操作失败, 消耗读CapacityUnit为", 0, "ErrorCode:", v1.ErrorCode, "ErrorMessage:", v1.ErrorMessage)
					} else {
						// 实际测试这里不会执行到
						fmt.Println("   --第", i1, "行操作失败, 消耗读CapacityUnit为", v1.Consumed.GetRead, "ErrorCode:", v1.ErrorCode, "ErrorMessage:", v1.ErrorMessage)
						consumed_write_total = consumed_write_total + v1.Consumed.GetRead()
					}
				}
			}
		}
		fmt.Printf("本次操作命中 %d 个, 失败 %d 个, 共消耗读CapacityUnit为 %d\n", succeed_total, failed_total, consumed_write_total)
	} else {
		fmt.Println("本次操作都失败，不消耗读CapacityUnit")
	}

	// get_range
	// 查询区间：[(1, INF_MIN), (4, INF_MAX))，左闭右开。
	inclusive_start_primary_key := &OTSPrimaryKey{
		"gid": 1,
		"uid": OTSColumnType_INF_MIN,
	}
	exclusive_end_primary_key := &OTSPrimaryKey{
		"gid": 4,
		"uid": OTSColumnType_INF_MAX,
	}
	columns_to_get = &OTSColumnsToGet{
		"gid", "uid", "name", "address", "mobile", "age",
	}

	// 选择方向
	// OTSDirection_FORWARD
	// OTSDirection_BACKWARD
	response_row_list, ots_err := ots_client.GetRange("myTable", OTSDirection_FORWARD,
		inclusive_start_primary_key, exclusive_end_primary_key, columns_to_get, 100)
	if ots_err != nil {
		fmt.Println(ots_err)
		os.Exit(1)
	}
	if response_row_list.GetRows() != nil {
		for i, v := range response_row_list.GetRows() {
			if v.GetPrimaryKeyColumns() != nil {
				fmt.Println("第 ", i, " 行:",
					"gid:", v.GetPrimaryKeyColumns().Get("gid"),
					"uid:", v.GetPrimaryKeyColumns().Get("uid"))
			} else {
				fmt.Println("第 ", i, " 行:")
			}

			fmt.Println("    - name信息:", v.GetAttributeColumns().Get("name"))
			fmt.Println("    - address信息:", v.GetAttributeColumns().Get("address"))
			fmt.Println("    - age信息:", v.GetAttributeColumns().Get("age"))
			fmt.Println("    - mobile信息:", v.GetAttributeColumns().Get("mobile"))
		}
	}
	fmt.Println("成功读取数据，消耗的读CapacityUnit为:", response_row_list.Consumed.GetRead())
	if response_row_list.GetNextStartPrimaryKey() != nil {
		fmt.Println("还有数据未读取完毕，用户可以继续调用GetRange()进行读取")
		fmt.Println("下次开始的主键:", response_row_list.GetNextStartPrimaryKey())
	}

	// delete_table
	ots_err = ots_client.DeleteTable("myTable")
	if ots_err != nil {
		fmt.Println(ots_err)
	}
	fmt.Println("测试完毕，表已删除")
}

File: doc/ots-python-sdk-2.0.8/ots2/example/interfaces.py
#!/bin/python
# -*- coding: utf8 -*-

import time
import logging
import unittest

from ots2 import *

ENDPOINT = 'your_instance_address'
ACCESSID = 'your_accessid'
ACCESSKEY = 'your_accesskey'
INSTANCENAME = 'your_instance_name'

def main():

    ots_client = OTSClient(ENDPOINT, ACCESSID, ACCESSKEY, INSTANCENAME)

    # delete_table
    try:
        ots_client.delete_table('myTable')
    except OTSServiceError,e:
        print e.get_error_code()
        print e.get_error_message()
    print u'表已删除。'
    
    # create_table
    # 注意：OTS是按设置的ReservedThroughput计量收费，即使没有读写也会产生费用。
    schema_of_primary_key = [('gid', 'INTEGER'), ('uid', 'INTEGER')]
    table_meta = TableMeta('myTable', schema_of_primary_key)
    reserved_throughput = ReservedThroughput(CapacityUnit(0, 0))
    ots_client.create_table(table_meta, reserved_throughput)
    print u'表已创建。'

    time.sleep(5)
    
    # list_table
    list_response = ots_client.list_table()
    print u'表的列表如下：'
    for table_name in list_response:
        print table_name
    
    # update_table
    try:
        # 由于刚创建表，需要2分钟之后才能调整表的预留读写吞吐量。
        # 注意：OTS是按设置的ReservedThroughput计量收费，即使没有读写也会产生费用。
        reserved_throughput = ReservedThroughput(CapacityUnit(0, 0))
        update_response = ots_client.update_table('myTable', reserved_throughput)
        print u'表的预留读吞吐量：%s' % update_response.reserved_throughput_details.capacity_unit.read
        print u'表的预留写吞吐量：%s' % update_response.reserved_throughput_details.capacity_unit.write
        print u'最后一次上调预留读写吞吐量时间：%s' % update_response.reserved_throughput_details.last_increase_time
        print u'最后一次下调预留读写吞吐量时间：%s' % update_response.reserved_throughput_details.last_decrease_time
        print u'UTC自然日内总的下调预留读写吞吐量次数：%s' % update_response.reserved_throughput_details.number_of_decreases_today
    except OTSServiceError,e:
        print e.get_error_code()
        print e.get_error_message()
    
    # describe_table
    describe_response = ots_client.describe_table('myTable')
    print u'表的名称: %s' % describe_response.table_meta.table_name
    print u'表的主键: %s' % describe_response.table_meta.schema_of_primary_key
    print u'表的预留读吞吐量：%s' % describe_response.reserved_throughput_details.capacity_unit.read
    print u'表的预留写吞吐量：%s' % describe_response.reserved_throughput_details.capacity_unit.write
    print u'最后一次上调预留读写吞吐量时间：%s' % describe_response.reserved_throughput_details.last_increase_time
    print u'最后一次下调预留读写吞吐量时间：%s' % describe_response.reserved_throughput_details.last_decrease_time
    print u'UTC自然日内总的下调预留读写吞吐量次数：%s' % describe_response.reserved_throughput_details.number_of_decreases_today
    
    # put_row
    primary_key = {'gid':1, 'uid':101}
    attribute_columns = {'name':'张三', 'mobile':111111111, 'address':'中国A地', 'age':20}
    condition = Condition('EXPECT_NOT_EXIST')
    consumed = ots_client.put_row('myTable', condition, primary_key, attribute_columns)
    print u'成功插入数据，消耗的写CapacityUnit为：%s' % consumed.write
    
    # get_row
    primary_key = {'gid':1, 'uid':101}
    columns_to_get = ['name', 'address', 'age']
    consumed, primary_key_columns, attribute_columns = ots_client.get_row('myTable', primary_key, columns_to_get)
    print u'成功读取数据，消耗的读CapacityUnit为：%s' % consumed.read
    print u'name信息：%s' % attribute_columns.get('name')
    print u'address信息：%s' % attribute_columns.get('address')
    print u'age信息：%s' % attribute_columns.get('age')

    # update_row
    primary_key = {'gid':1, 'uid':101}
    update_of_attribute_columns = {
        'put' : {'name':'张三丰', 'address':'中国B地'},
        'delete' : ['mobile', 'age'],
    }
    condition = Condition('EXPECT_EXIST')
    consumed = ots_client.update_row('myTable', condition, primary_key, update_of_attribute_columns) 
    print u'成功更新数据，消耗的写CapacityUnit为：%s' % consumed.write
    
    # delete_row
    primary_key = {'gid':1, 'uid':101}
    condition = Condition('IGNORE')
    consumed = ots_client.delete_row('myTable', condition, primary_key) 
    print u'成功删除数据，消耗的写CapacityUnit为：%s' % consumed.write
    
    # batch_write_row
    primary_key = {'gid':2, 'uid':202}
    attribute_columns = {'name':'李四', 'address':'中国某地', 'age':20}
    condition = Condition('EXPECT_NOT_EXIST')
    put_row_item = PutRowItem(condition, primary_key, attribute_columns)
    
    primary_key = {'gid':3, 'uid':303}
    condition = Condition('IGNORE')
    update_of_attribute_columns = {
        'put' : {'name':'张三', 'address':'中国某地'},
        'delete' : ['mobile', 'age'],
    }
    update_row_item = UpdateRowItem(condition, primary_key, update_of_attribute_columns)
    
    primary_key = {'gid':4, 'uid':404}
    condition = Condition('IGNORE')
    delete_row_item = DeleteRowItem(condition, primary_key)
    
    table_item1  = {'table_name':'myTable', 'put':[put_row_item], 'update':[update_row_item], 'delete':[delete_row_item]}
    table_item2  = {'table_name':'notExistTable', 'put':[put_row_item], 'update':[update_row_item], 'delete':[delete_row_item]}
    batch_list = [table_item1, table_item2]
    batch_write_response = ots_client.batch_write_row(batch_list) 

    # 每一行操作都是独立的，需要分别判断是否成功。对于失败子操作进行重试。
    retry_count = 0
    operation_list = ['put', 'update', 'delete']
    while retry_count < 3:
        failed_batch_list = []
        for i in range(len(batch_write_response)):
            table_item = batch_write_response[i]
            for operation in operation_list:
                operation_item = table_item.get(operation)
                if not operation_item:
                    continue
                print u'操作：%s' % operation
                for j in range(len(operation_item)):
                    row_item = operation_item[j]
                    print u'操作是否成功：%s' % row_item.is_ok
                    if not row_item.is_ok:
                        print u'错误码：%s' % row_item.error_code
                        print u'错误信息：%s' % row_item.error_message
                        add_batch_write_item(failed_batch_list, batch_list[i]['table_name'], operation, batch_list[i][operation][j])
                    else:
                        print u'本次操作消耗的写CapacityUnit为：%s' % row_item.consumed.write

        if not failed_batch_list:
            break
        retry_count += 1
        batch_list = failed_batch_list
        batch_write_response = ots_client.batch_write_row(batch_list)
    
    # batch_get_row
    row1_primary_key = {'gid':1, 'uid':101}
    row2_primary_key = {'gid':2, 'uid':202}
    row3_primary_key = {'gid':3, 'uid':303}
    columns_to_get = ['name', 'address', 'mobile', 'age']
    batch_list = [('myTable', [row1_primary_key, row2_primary_key, row3_primary_key], columns_to_get)]
    batch_list = [('notExistTable', [row1_primary_key, row2_primary_key, row3_primary_key], columns_to_get)]
    batch_get_response = ots_client.batch_get_row(batch_list) 

    # 每一行操作都是独立的，需要分别判断是否成功。对于失败子操作进行重试。
    retry_count = 0
    while retry_count < 3:
        failed_batch_list = []
        for i in range(len(batch_get_response)):
            table_item = batch_get_response[i]
            for j in range(len(table_item)):
                row_item = table_item[j]
                print u'操作是否成功：%s' % row_item.is_ok
                if not row_item.is_ok:
                    print u'错误码：%s' % row_item.error_code
                    print u'错误信息：%s' % row_item.error_message
                    add_batch_get_item(failed_batch_list, batch_list[i][0], batch_list[i][1][j], batch_list[i][2])
                else:
                    print u'name信息：%s' % row_item.attribute_columns.get('name')
                    print u'address信息：%s' % row_item.attribute_columns.get('address')
                    print u'mobile信息：%s' % row_item.attribute_columns.get('mobile')
                    print u'age信息：%s' % row_item.attribute_columns.get('age')
                    print u'本次操作消耗的读CapacityUnit为：%s' % row_item.consumed.read

        if not failed_batch_list:
            break
        retry_count += 1
        batch_list = failed_batch_list
        batch_get_response = ots_client.batch_get_row(batch_list)
    
    # get_range
    # 查询区间：[(1, INF_MIN), (4, INF_MAX))，左闭右开。
    inclusive_start_primary_key = {'gid':1, 'uid':INF_MIN} 
    exclusive_end_primary_key = {'gid':4, 'uid':INF_MAX} 
    columns_to_get = ['name', 'address', 'mobile', 'age']
    consumed, next_start_primary_key, row_list = ots_client.get_range(
                'myTable', 'FORWARD', 
                inclusive_start_primary_key, exclusive_end_primary_key,
                columns_to_get, 100
    )
    for row in row_list:
        attribute_columns = row[1]
        print u'name信息为：%s' % attribute_columns.get('name')
        print u'address信息为：%s' % attribute_columns.get('address')
        print u'mobile信息为：%s' % attribute_columns.get('mobile')
        print u'age信息为：%s' % attribute_columns.get('age')
    print u'本次操作消耗的读CapacityUnit为：%s' % consumed.read
    print u'下次开始的主键：%s' % next_start_primary_key

    # xget_range
    consumed_counter = CapacityUnit(0, 0)
    inclusive_start_primary_key = {'gid':1, 'uid':INF_MIN} 
    exclusive_end_primary_key = {'gid':4, 'uid':INF_MAX} 
    columns_to_get = ['gid', 'uid', 'name', 'address', 'mobile', 'age']
    range_iter = ots_client.xget_range(
                'myTable', 'FORWARD', 
                inclusive_start_primary_key, exclusive_end_primary_key,
                consumed_counter, columns_to_get, 100
    )
    for (primary_key_columns, attribute_columns) in range_iter:
        print u'gid信息为：%s' % primary_key_columns.get('gid')
        print u'uid信息为：%s' % primary_key_columns.get('uid')
        print u'name信息为：%s' % attribute_columns.get('name')
        print u'address信息为：%s' % attribute_columns.get('address')
        print u'mobile信息为：%s' % attribute_columns.get('mobile')
        print u'age信息为：%s' % attribute_columns.get('age')

def add_batch_write_item(batch_list, table_name, operation, item):
    for table_item in batch_list:
        if table_item.get('table_name') == table_name:
            operation_item = table_item.get(operation)
            if not operation_item:
                table_item[operation] = [item]
            else:
                operation_item.append(item)
            return
    # not found
    table_item = {'table_name':table_name, operation:[item]}
    batch_list.append(table_item)

def add_batch_get_item(batch_list, table_name, item, columns_to_get):
    for table_item in batch_list:
        if table_item[0] == table_name:
            row_item_list = table_item[1]
            row_item_list.append(item)
            return
    # not found
    table_item = (table_name, [item], columns_to_get)
    batch_list.append(table_item)
                
if __name__ == '__main__':
    main()
